File: app/src/main/java/net/radityalabs/contactapp/presentation/ui/animation/CircularAnimUtil.java
package net.radityalabs.contactapp.presentation.ui.animation;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.ViewAnimationUtils;
import android.view.ViewGroup;
import android.widget.ImageView;

/**
 * Created by radityagumay on 4/13/17.
 */

public class CircularAnimUtil {

    public static final long PERFECT_MILLS = 618;
    public static final int MINI_RADIUS = 0;

    @SuppressLint("NewApi")
    public static void show(View myView, float startRadius, long durationMills) {
        if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.LOLLIPOP) {
            myView.setVisibility(View.VISIBLE);
            return;
        }
        int cx = (myView.getLeft() + myView.getRight()) / 2;
        int cy = (myView.getTop() + myView.getBottom()) / 2;
        int w = myView.getWidth();
        int h = myView.getHeight();
        int finalRadius = (int) Math.sqrt(w * w + h * h) + 1;
        Animator anim = ViewAnimationUtils.createCircularReveal(myView, cx, cy, startRadius, finalRadius);
        myView.setVisibility(View.VISIBLE);
        anim.setDuration(durationMills);
        anim.start();
    }

    @SuppressLint("NewApi")
    public static void hide(final View myView, float endRadius, long durationMills) {
        if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.LOLLIPOP) {
            myView.setVisibility(View.INVISIBLE);
            return;
        }
        int cx = (myView.getLeft() + myView.getRight()) / 2;
        int cy = (myView.getTop() + myView.getBottom()) / 2;
        int w = myView.getWidth();
        int h = myView.getHeight();
        int initialRadius = (int) Math.sqrt(w * w + h * h) + 1;
        Animator anim = ViewAnimationUtils.createCircularReveal(myView, cx, cy, initialRadius, endRadius);
        anim.setDuration(durationMills);
        anim.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                super.onAnimationEnd(animation);
                myView.setVisibility(View.INVISIBLE);
            }
        });

        anim.start();
    }

    @SuppressLint("NewApi")
    public static void startActivityForResult(final Activity thisActivity, final Intent intent, final Integer requestCode, final Bundle bundle, final View triggerView, int colorOrImageRes, long durationMills) {
        if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.LOLLIPOP) {
            thisActivity.startActivity(intent);
            return;
        }
        int[] location = new int[2];
        triggerView.getLocationInWindow(location);
        final int cx = location[0] + triggerView.getWidth() / 2;
        final int cy = location[1] + triggerView.getHeight() / 2;
        final ImageView view = new ImageView(thisActivity);
        view.setScaleType(ImageView.ScaleType.CENTER_CROP);
        view.setImageResource(colorOrImageRes);
        final ViewGroup decorView = (ViewGroup) thisActivity.getWindow().getDecorView();
        int w = decorView.getWidth();
        int h = decorView.getHeight();
        decorView.addView(view, w, h);

        int maxW = Math.max(cx, w - cx);
        int maxH = Math.max(cy, h - cy);
        final int finalRadius = (int) Math.sqrt(maxW * maxW + maxH * maxH) + 1;
        Animator anim = ViewAnimationUtils.createCircularReveal(view, cx, cy, 0, finalRadius);
        int maxRadius = (int) Math.sqrt(w * w + h * h) + 1;
        if (durationMills == PERFECT_MILLS) {
            double rate = 1d * finalRadius / maxRadius;
            durationMills = (long) (PERFECT_MILLS * rate);
        }
        final long finalDuration = durationMills;
        anim.setDuration(finalDuration);
        anim.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                super.onAnimationEnd(animation);
                if (requestCode == null)
                    thisActivity.startActivity(intent);
                else if (bundle == null)
                    thisActivity.startActivityForResult(intent, requestCode);
                else
                    thisActivity.startActivityForResult(intent, requestCode, bundle);
                thisActivity.overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
                triggerView.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        Animator anim = ViewAnimationUtils.createCircularReveal(view, cx, cy, finalRadius, 0);
                        anim.setDuration(finalDuration);
                        anim.addListener(new AnimatorListenerAdapter() {
                            @Override
                            public void onAnimationEnd(Animator animation) {
                                super.onAnimationEnd(animation);
                                try {
                                    decorView.removeView(view);
                                } catch (Exception e) {
                                    e.printStackTrace();
                                }
                            }
                        });
                        anim.start();
                    }
                }, 1000);

            }
        });
        anim.start();
    }

    public static void startActivityForResult(Activity thisActivity, Intent intent, Integer requestCode, View triggerView, int colorOrImageRes) {
        startActivityForResult(thisActivity, intent, requestCode, null, triggerView, colorOrImageRes, PERFECT_MILLS);
    }

    public static void startActivity(Activity thisActivity, Intent intent, View triggerView, int colorOrImageRes, long durationMills) {
        startActivityForResult(thisActivity, intent, null, null, triggerView, colorOrImageRes, durationMills);
    }

    public static void startActivity(Activity thisActivity, Intent intent, View triggerView, int colorOrImageRes) {
        startActivity(thisActivity, intent, triggerView, colorOrImageRes, PERFECT_MILLS);
    }

    public static void startActivity(Activity thisActivity, Class<?> targetClass, View triggerView, int colorOrImageRes) {
        startActivity(thisActivity, new Intent(thisActivity, targetClass), triggerView, colorOrImageRes, PERFECT_MILLS);
    }

    public static void show(View myView) {
        show(myView, MINI_RADIUS, PERFECT_MILLS);
    }

    public static void hide(View myView) {
        hide(myView, MINI_RADIUS, PERFECT_MILLS);
    }

}

File: app/src/main/java/net/radityalabs/contactapp/presentation/ui/animation/CircularAnimUtil.kt
package net.radityalabs.contactapp.presentation.ui.animation

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.ViewAnimationUtils
import android.view.ViewGroup
import android.widget.ImageView

/**
 * Created by radityagumay on 4/13/17.
 */

object CircularAnimUtil {

    val PERFECT_MILLS: Long = 618
    val MINI_RADIUS = 0

    @SuppressLint("NewApi")
    fun show(myView: View, startRadius: Float, durationMills: Long) {
        if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.LOLLIPOP) {
            myView.visibility = View.VISIBLE
            return
        }
        val cx = (myView.left + myView.right) / 2
        val cy = (myView.top + myView.bottom) / 2
        val w = myView.width
        val h = myView.height
        val finalRadius = Math.sqrt((w * w + h * h).toDouble()).toInt() + 1
        val anim = ViewAnimationUtils.createCircularReveal(myView, cx, cy, startRadius, finalRadius.toFloat())
        myView.visibility = View.VISIBLE
        anim.duration = durationMills
        anim.start()
    }

    @SuppressLint("NewApi")
    fun hide(myView: View, endRadius: Float, durationMills: Long) {
        if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.LOLLIPOP) {
            myView.visibility = View.INVISIBLE
            return
        }
        val cx = (myView.left + myView.right) / 2
        val cy = (myView.top + myView.bottom) / 2
        val w = myView.width
        val h = myView.height
        val initialRadius = Math.sqrt((w * w + h * h).toDouble()).toInt() + 1
        val anim = ViewAnimationUtils.createCircularReveal(myView, cx, cy, initialRadius.toFloat(), endRadius)
        anim.duration = durationMills
        anim.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                super.onAnimationEnd(animation)
                myView.visibility = View.INVISIBLE
            }
        })

        anim.start()
    }

    @SuppressLint("NewApi")
    fun startActivityForResult(thisActivity: Activity, intent: Intent, requestCode: Int?, bundle: Bundle?, triggerView: View, colorOrImageRes: Int, durationMills: Long) {
        var durationMills = durationMills
        if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.LOLLIPOP) {
            thisActivity.startActivity(intent)
            return
        }
        val location = IntArray(2)
        triggerView.getLocationInWindow(location)
        val cx = location[0] + triggerView.width / 2
        val cy = location[1] + triggerView.height / 2
        val view = ImageView(thisActivity)
        view.scaleType = ImageView.ScaleType.CENTER_CROP
        view.setImageResource(colorOrImageRes)
        val decorView = thisActivity.window.decorView as ViewGroup
        val w = decorView.width
        val h = decorView.height
        decorView.addView(view, w, h)

        val maxW = Math.max(cx, w - cx)
        val maxH = Math.max(cy, h - cy)
        val finalRadius = Math.sqrt((maxW * maxW + maxH * maxH).toDouble()).toInt() + 1
        val anim = ViewAnimationUtils.createCircularReveal(view, cx, cy, 0f, finalRadius.toFloat())
        val maxRadius = Math.sqrt((w * w + h * h).toDouble()).toInt() + 1
        if (durationMills == PERFECT_MILLS) {
            val rate = 1.0 * finalRadius / maxRadius
            durationMills = (PERFECT_MILLS * rate).toLong()
        }
        val finalDuration = durationMills
        anim.duration = finalDuration
        anim.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                super.onAnimationEnd(animation)
                if (requestCode == null)
                    thisActivity.startActivity(intent)
                else if (bundle == null)
                    thisActivity.startActivityForResult(intent, requestCode)
                else
                    thisActivity.startActivityForResult(intent, requestCode, bundle)
                thisActivity.overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
                triggerView.postDelayed({
                    val anim = ViewAnimationUtils.createCircularReveal(view, cx, cy, finalRadius.toFloat(), 0f)
                    anim.duration = finalDuration
                    anim.addListener(object : AnimatorListenerAdapter() {
                        override fun onAnimationEnd(animation: Animator) {
                            super.onAnimationEnd(animation)
                            try {
                                decorView.removeView(view)
                            } catch (e: Exception) {
                                e.printStackTrace()
                            }

                        }
                    })
                    anim.start()
                }, 1000)

            }
        })
        anim.start()
    }

    fun startActivityForResult(thisActivity: Activity, intent: Intent, requestCode: Int?, triggerView: View, colorOrImageRes: Int) {
        startActivityForResult(thisActivity, intent, requestCode, null, triggerView, colorOrImageRes, PERFECT_MILLS)
    }

    @JvmOverloads fun startActivity(thisActivity: Activity, intent: Intent, triggerView: View, colorOrImageRes: Int, durationMills: Long = PERFECT_MILLS) {
        startActivityForResult(thisActivity, intent, null, null, triggerView, colorOrImageRes, durationMills)
    }

    fun startActivity(thisActivity: Activity, targetClass: Class<*>, triggerView: View, colorOrImageRes: Int) {
        startActivity(thisActivity, Intent(thisActivity, targetClass), triggerView, colorOrImageRes, PERFECT_MILLS)
    }

    fun show(myView: View) {
        show(myView, MINI_RADIUS.toFloat(), PERFECT_MILLS)
    }

    fun hide(myView: View) {
        hide(myView, MINI_RADIUS.toFloat(), PERFECT_MILLS)
    }

}
